### PR TITLE
fix: typed da classing

### PR DESCRIPTION
**Pull request by Jackmin801**
Goals:

- Addresses #1041 
- This solution is different from the one suggested by @JohannesMessner in https://github.com/docarray/docarray/pull/991/
  - This solution is cleaner imo

### CODE REVIEW

**Comment by JohannesMessner** [comment]
So if I understand correctly, this basically caches the parametrized classes so that they will not be recreated when it is not necessaray, right? Are there any scenarios where we could end up with two class "instances" with two different caches, such that the original problem could still occur? Maybe through multi-threading/processing? Not sure myself...

**Comment by Jackmin801** [comment]
I dont think worker processes can return a class that was defined in the worker process anyways. So even if different instances existed, it would be illegal for them to interact.

File: docarray/array/abstract_array.py [hunk 1]
```diff
@@ -16,35 +26,39 @@
 class AnyDocumentArray(Sequence[BaseDocument], Generic[T_doc], AbstractType):
     document_type: Type[BaseDocument]
     tensor_type: Type['AbstractTensor'] = NdArray
+    __typed_da__: Dict[Type[BaseDocument], Type] = {}
```
Comment (samsja) [inline docarray/array/abstract_array.py:29]:
```suggestion
    __typed_da__: Dict[Type[BaseDocument], Type] = dict()
```

**Review by samsja — APPROVED** [review]
neat

**Review by samsja — CHANGES_REQUESTED** [review]
oups I forgot that I ask for a change

File: docarray/array/abstract_array.py (hunk 1 above)
Comment (Jackmin801) [reply docarray/array/abstract_array.py:29]: Is there an error case with using `{}`? I thought it was faster by a few milliseconds?

Comment (samsja) [reply docarray/array/abstract_array.py]: Screenshot from 2023-01-25 11-52-36

it is not safe to use {} or [] inside function argument default value

Comment (Jackmin801) [reply docarray/array/abstract_array.py]: But its a static variable... its not a default arg

**Review by samsja — APPROVED** [review]

**Pull request merged**